mtags/src/main/scala/scala/meta/internal/mtags/GlobalSymbolIndex.scala

```diff
@@ -0,0 +1,84 @@
+ */
+trait GlobalSymbolIndex {
+
+  /**
+   * Lookup the definition of a global symbol.
+   *
+   * Returns the path of the file that defines the symbol but does not include
+   * the exact position of the definition. Computing the range position of the
+   * definition is left for the client and can be done using the mtags module.
+   *
+   * @param symbol a global SemanticDB symbol. For comprehensive documentation
+   *               of how a symbol is formatted consule the specification:
+   *               https://scalameta.org/docs/semanticdb/specification.html#symbol
+   *
+   *  Examples: {{{
+   *    "scala/Option#"                 // Option class
+   *    "scala/Option."                 // Option companion object
+   *    "scala/Option#.get()."          // `get` method for Option
+   *    "scala/Predef.String#"          // String type alias in Predef
+   *    "java/lang/String#format(+1)."  // Static `format` method for strings
+   *    "java/util/Map#Entry#"          // Static inner interface.
+   *  }}}
+   * @return the definition of the symbol, if any.
+   */
+  def definition(symbol: String): Option[SymbolDefinition]
+
+  /**
+   * Add an individual Java or Scala source file to the index.
+   *
+   * @param file the absolute path to the source file, can be a path
+   *            on disk or inside of a jar/zip file.
+   * @param sourceDirectory the enclosing project source directory if
+   *                        file is on disk, used to relativize `file`.
+   *                        Can be None if file is inside a zip file
+   *                        assuming the file path is already relativized
+   *                        by that point.
+   */
+  def addSourceFile(
+      file: AbsolutePath,
+      sourceDirectory: Option[AbsolutePath]
+  ): Unit
+
+  /**
+   * Index a jar or zip file containing Scala and Java source files.
+   *
+   * Published artifacts typically have accompanying sources.jar files that
+   * include the project sources.
+   *
+   * Sources of a published library can be fetched from the command-line with
+   * coursier using the --sources flag:
+   * {{{
+   *   $ coursier fetch com.thoughtworks.qdox:qdox:2.0-M9 --sources
+   *   $HOME/.coursier/cache/v1/https/repo1.maven.org/maven2/com/thoughtworks/qdox/qdox/2.0-M9/qdox-2.0-M9-sources.jar.coursier/cache/v1/https/repo1.maven.org/maven2/com/thoughtworks/qdox/qdox/2.0-M9/qdox-2.0-M9-sources.jar
+   * }}}
+   *
+   * Sources can be fetched from an sbt build through the `updateClassifiers` task:
+   * {{{
+   *   val sourceJars = for {
+   *     configurationReport <- updateClassifiers.in(input).value.configurations
+   *     moduleReport <- configurationReport.modules
+   *     (artifact, file) <- moduleReport.artifacts
+   *     if artifact.classifier.contains("sources")
+   *   } yield file
+   * }}}
+   *
+   * @param jar the path to a single jar or zip file.
+   */
+  def addSourceJar(jar: AbsolutePath): Unit
```

Comment:
Same comment as above about returning `Unit` vs a more expressive error-aware type

Reply:
It's a good question but I'm still not sure what the best solution  is